src/main/java/com/exactpro/th2/taskutils/FutureTracker.java

```diff
@@ -98,27 +209,81 @@ public void awaitRemaining (long timeoutMillis) {
                     }
                 }
             } catch (ExecutionException | TimeoutException e) {
+                // do nothing
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
             } catch (Exception e) {
                 throw new RuntimeException(e);
             }
         }
+        if (Thread.currentThread().isInterrupted()) {
+            throw new InterruptedException();
+        }
     }
 
     /**
      * Informational method for getting remaining number of futures
+     *
      * @return number of unfinished futures
      */
-    public int remaining () {
-        synchronized (futures) {
+    public int remaining() {
+        lock.readLock().lock();
+        try {
             return futures.size();
+        } finally {
+            lock.readLock().unlock();
         }
     }
 
+    /**
+     * Limit of trackable features if positive value, otherwise tracker unlimited
+     *
+     * @return max number of trackable futures if positive value, otherwise tracker unlimited
+     */
+    public int limit() {
+        return limit;
+    }
+
+    /**
+     * Has tracker got limit for number of trackable futures
+     *
+     * @return true if {@link #limit()} is positive, otherwise false
+     */
+    public boolean hasLimit() {
+        return hasLimit;
+    }
+
     /**
      * Informs if there are any active futures
+     *
      * @return true if there are no active futures
      */
-    public boolean isEmpty () {
+    public boolean isEmpty() {
         return remaining() == 0;
     }
+
+    /**
+     * Informs if future tracker can accept new futures or not
+     *
+     * @return current state of {@link FutureTracker} instance
+     */
+    public boolean isEnabled() {
+        return enabled;
+    }
+
+    private void addFuture(CompletableFuture<T> future) {
+        if (!enabled) {
+            throw new IllegalStateException("Future tracker is disabled before tracking start");
```

Comment:
I don't think the last part is needed here. All user needs to know is that the tracking does not work
```suggestion
            throw new IllegalStateException("future tracker is disabled");
```